### ISSUE TITLE
DB CASE WHEN in SELECT

### ISSUE THREAD

**Issue by mariuszkrzaczkowski** [yiisoft/yii2#12781]
New feature, adding the possibility to use CASE WHEN in yii\db\Query

## DB support for CASE WHEN

- [ ] MySQL
- [ ] PostgreSQL
- [ ] MSSQL
- [ ] Oracle
- [ ] SQLite
- [ ] Cubrid

## TODO

- [ ] Choose syntax for it
- [ ]  Write tests
- [ ] Implement

**Comment by Alex-Code** [comment]
This is already possible using an `Expression`
```php
->select(['c' => new \yii\db\Expression("CASE WHEN 1>0 THEN 'true' ELSE 'false' END")])->
```

**Comment by samdark** [comment]
It's [part of SQL99](https://mariadb.com/kb/en/sql-99/sqls-case-expression/) so likely to supported by all DBs we have. A shortcut could be introduced...

**Comment by mariuszkrzaczkowski** [comment]
Thank you very much

**Comment by dynasource** [comment]
@samdark, how do you think this shortcut should look like?

**Comment by samdark** [comment]
```php
$c = new \yii\db\Case()->when("1>0")->then(true)->else(false);
$query->select(['c' => $c])->...
```

**Comment by rob006** [comment]
```php
$query->select([
	'c' => (new \yii\db\Case())
			->when("c > 0")->then(1)
			->when("c < 0")->then(-1)
			->else(0),
]);
```
```php
$query->select([
	'c' => (new \yii\db\Case('c'))
			->when("a")->then(1)
			->when("b")->then(-1)
			->else(0),
]);
```


?

Also, `case` is reserved keyword.

**Comment by samdark** [comment]
@rob006 what's constructor argument in the 2nd argument for?

Yeah, can't use `Case` :( Need another name. Probably `CaseQuery` would do.

**Comment by rob006** [comment]
> what's constructor argument in the 2nd argument for?

To cover:
```sql
SELECT CASE c WHEN 'a' THEN -1 WHEN 'b' THEN -1 ELSE 0 END ...
```

This probably will require some escaping...

**Comment by samdark** [comment]
Oh... It could be multi-level.

**Comment by mariuszkrzaczkowski** [comment]
I like that. Can this query be supported?
CASE WHEN lead.status IS NULL || lead.status = "" THEN "" ELSE lead.status END AS leadstatus

**Comment by samdark** [comment]
@mariuszkrzaczkowski 

```php
$case = new \yii\db\Case())
    ->when('lead.status IS NULL || lead.status = ""')->then('')
    ->else('lead.status');

$query->select([
    'leadstatus' =>$case,
]);
```

**Comment by Alex-Code** [comment]
Probably need support for params too
Either in the method or as a separate one.
```php
$case = new \yii\db\Case()
    ->when('lead.status IS NULL || lead.status = :param', [':param' => 1234])->then('')
    ->else('lead.status');

$case = new \yii\db\Case()
    ->when('lead.status IS NULL || lead.status = :param')->then('')
    ->else('lead.status')
    ->params([':param' => 1234]);
```

**Comment by samdark** [comment]
Parameters are already handled on query level so I don't think there's a need to add special support at `CaseQuery` level.

**Comment by Alex-Code** [comment]
That's true, but it would keep it self contained and reusable in more than one query, similar to how `Expression` is.

**Comment by rob006** [comment]
> Parameters are already handled on query level so I don't think there's a need to add special support at CaseQuery level.

We could say the same about `yii\db\Expression`. And most of `yii\db\Query` methods.

**Comment by cebe** [comment]
> Yeah, can't use Case :( Need another name. Probably CaseQuery would do.

`CaseExpression`

**Comment by samdark** [comment]
@cebe yes, a good name.

**Comment by mariuszkrzaczkowski** [comment]
Where is this file?

**Comment by samdark** [comment]
Which file?

**Comment by mariuszkrzaczkowski** [comment]
\yii\db\Case

**Comment by samdark** [comment]
In our heads. We're just thinking loud.

**Comment by mariuszkrzaczkowski** [comment]
Ok I understand good ideas.
Case should work in OrderBy

**Comment by samdark** [comment]
No only in orderBy. It should be everywhere.

**Comment by mariuszkrzaczkowski** [comment]
You also have to predict many conditions
```sql
CASE 
          WHEN org.size IN (0, 1) THEN '<26'
          WHEN org.size = 2 THEN '26-50'
          WHEN org.size = 3 THEN '51-100'
          WHEN org.size = 4 THEN '101-250'
          WHEN org.size = 5 THEN '251-500'
          WHEN org.size = 6 THEN '501-1000'
          WHEN org.size = 7 THEN '1001-5000'
        ELSE '5000+' END AS Size, ....
```

**Comment by samdark** [comment]
```php
$case = new \yii\db\CaseExpression())
    ->when(['org.size' => [0, 1]])->then('<26')
   ->when(['org.size' => 2])->then('26-50')
   ->when(['org.size' => 3])->then('51-100')
   ->when(['org.size' => 4])->then('101-250')
   ->when(['org.size' => 5])->then('251-500')
   ->when(['org.size' => 6])->then('501-1000')
   ->when(['org.size' => 7])->then('1001-5000')
    ->else('5000+');

$query->select([
    'Size' =>$case,
]);
```

**Comment by Vovan-VE** [comment]
There are mix of `'plain value'` and `'table.column'` in an examples above. Probably, `then()` and `else()` needs pair methods like `thenColumn()` and `elseColumn()`.

Also, PHP < 7 does not allow method `else()`.

**Comment by samdark** [comment]
> Also, PHP < 7 does not allow method else().

`orElse`?

**Comment by SilverFire** [comment]
I'd prefer `default`

**Comment by vercotux** [comment]
>I'd prefer `default`

`default` is also a [reserved keyword](https://secure.php.net/manual/en/reserved.keywords.php) in PHP < 7.0.0

**Comment by vercotux** [comment]
Unfortunately `default` was also a reserved keyword in PHP<7.0. But, `when()` and `then()` should not be separate methods anyway since you would never call one without the other, would you? How about prefixing the methods with `add`, since that's pretty much what you're doing (adding conditions):

    $case = (new CaseExpression())
        ->addWhen(['org.size' => [0,1]], '<26')
        ->addWhen(['org.size' => 2], '26-50', $optionalParams = [])
        ->addWhen(['org.size' => 3], '51-100')
        ->addElse('100+');

It would produce:

    CASE
        WHEN org.size IN (0, 1) THEN '<26'
        WHEN org.size = 2 THEN '26-50'
        WHEN org.size = 3 THEN '51-100'
        ELSE '100+'
    END

And for the abbreviated "simple case" syntax:

    $case = (new CaseExpression('org.size'))
        ->addWhen(1, '<26')
        ->addWhen(2, '26-50')
        ->addWhen(3, '51-100')
        ->addElse('100+');

It would produce:

    CASE org.size
        WHEN 1 THEN '<26'
        WHEN 2 THEN '26-50'
        WHEN 3 THEN '51-100'
        ELSE '100+'
    END

@samdark I wonder how you would build those array-based conditions inside CaseExpression though? It would require access to `yii\db\Connection` which is not available within CaseExpression assuming it extends from `yii\db\Expression`.

**Comment by samdark** [comment]
@SilverFire is it possible to implement such a thing with your recent proposal?

**Comment by bologer** [comment]
@samdark guys, is there any progress on this? What is the best way so far? Using new `yii\db\Expression`?

**Comment by cebe** [comment]
> Using new `yii\db\Expression`?

yes
